### ISSUE TITLE
3.4.0 - Error using map widgets - widget configuration screen is not displayed 

### ISSUE THREAD

**Issue by kastayou** [thingsboard/thingsboard#6970]
**Describe the bug**

After migration from Thingsboard v3.3.4.1 existing image map widget present in my dashboard shows the following error:

![image](https://user-images.githubusercontent.com/30028174/179846534-b650285a-bb2d-4013-8f05-3221ef64925b.png)

If I try to recreate the widget by adding a new map image widget, the widget configuration window is not displayed and I have the following error message in the browser console:

```
errorMessages: Array(1)
0: "Failed to compile widget script. ↵ Error: TbMapWidgetV2.settingsSchema is not a function"
length: 1
__proto__: Array(0)
widgetInfo:
alias: "error"
componentFactory: Hg {componentDef: {…}, ngModule: zE, componentType: ƒ, selector: "ng-component", ngContentSelectors: Array(0), …}
controllerScript: "self.onInit = function() {}"
dataKeySettingsSchema: "{}↵"
defaultConfig: "{↵"title": "Widget failed to load",↵"datasources": [],↵"settings": {}↵}↵"
description: null
image: null
resources: []
settingsSchema: "{}↵"
sizeX: 8
sizeY: 6
templateCss: ""
templateHtml: "<div class="tb-widget-error-container"><div translate class="tb-widget-error-msg">widget.widget-type-load-error</div><div *ngFor="let error of errorMessages" class="tb-widget-error-msg">{{ error }}</div></div>"
type: "latest"
typeParameters: {}
widgetName: "Error loading widget"
```

The same error occurs if I try to add an image map, openstreet map or google map widget.

**Your Server Environment**
* own setup
  * local infrastructure
  * ThingsBoard Version 3.4.0 (upgraded from v3.3.4.1
  * Red Hat Enterprise Linux release 8.4 (Ootpa)

**Your Client Environment**
**Desktop (please complete the following information):**
Windows 10
Same error with the following browsers:
Chrome Version 73.0.3683.86 (Official Build) (64-bit)
Edge Version 100.0.1185.50

**To Reproduce**
Create a new dashboard
Open the new dashboard and clicke on create a new widget
Select "Maps" and "Openstreetmap"

**Expected behavior**
The configuration window of the widget is displayed to configure the widget and its datasource.
But the configuration window is not displayed and an error message is shown in the browser console.

**Comment by kastayou** [comment]
I have the same issue with "timeseries line charts" widget with the following error message:

```
Widget wasn't loaded due to the following errors:
Failed to compile widget script. Error: TbFlot.settingsSchema is not a function
```

**Comment by Backdraft007** [comment]
Maybe a browser cache problem? Can you clean the cache with CTRL+F5.

**Comment by mde2017** [comment]
Hi @kastayou,

had the same issue, but was able to resolve this problem.

I guess you are using CE version?
There is currently a bug mentioned here: 
https://github.com/thingsboard/thingsboard/issues/6971

After making the mentioned changes for the workaround, it also seems to be neccessary to start the upgrade script again via:
`sudo /usr/share/thingsboard/bin/install/upgrade.sh --fromVersion=3.3.4`

Then start the ThingsBoard Service and the Widgets should be displayed correctly again (I checked both maps and timeseries line charts)

**Comment by kastayou** [comment]
> Maybe a browser cache problem? Can you clean the cache with CTRL+F5.

No, it did not work

**Comment by kastayou** [comment]
I have restarted from my thingsboard CE 3.3.4.1.
I have downloaded again the rpm package since I have seen that it has been updated recently.
I have performed the upgrade from 3.3.4.1.
No error during the upgrqde

After the upgrade, my map widgets are correctly displayed and I am able to add a new map widget.

Problem is solved

**Comment by kastayou** [comment]
> Hi @kastayou,
> 
> had the same issue, but was able to resolve this problem.
> 
> I guess you are using CE version? There is currently a bug mentioned here: #6971
> 
> After making the mentioned changes for the workaround, it also seems to be neccessary to start the upgrade script again via: `sudo /usr/share/thingsboard/bin/install/upgrade.sh --fromVersion=3.3.4`
> 
> Then start the ThingsBoard Service and the Widgets should be displayed correctly again (I checked both maps and timeseries line charts)

Thanks, for your feedback. yes I think it was the same issue. It looks like the upgrade script has been fixed since I did not reproduce the issue using the latest rpm package.

**Comment by mgogh** [comment]
Hi,

I encounter the same problem (I am in docker installation) and I try to update TB PE to 3.4 (I was in 3.3.4.1), where can I find the "upgrade script" that you mentionned ?

I have the same error and the support tell me that method "TbFlot.settingsSchema" was deleted...

With your fix, you can still use this method ?

Regards

**Comment by mde2017** [comment]
Hi @mgogh,

I'm not using ThingsBoard with Docker currently, but there is a different upgrade process for docker:
https://thingsboard.io/docs/user-guide/install/docker/#upgrading

**Comment by Eric-Ruud** [comment]
@mgogh did you ever resolve this? I'm experiencing the same problem in 3.4.1

**Comment by Eric-Ruud** [comment]
Also @mgogh are you seeing this in a custom widget or in the packaged widgets?

**Comment by mgogh** [comment]
@Eric-Ruud, I resolve it in redevelopping my custom widgets and trying to recreate the correct "settings schema" but unfortunately there is a huge gap between the TB documentation about JSON schema and the actual version of JSON schema.

**Comment by evanderbird** [comment]
Hi @mgogh, how you resolve it? I experienced this on PE version 3.4 too,
can you tell me what did you do , what script you edit to fix this?

**Comment by Vandrovec-SEA** [comment]
We need to migrate quite soon to the latest version, as our old server will be gone in a month. I'm getting the same errors as OP with my custom flot component.

I found that the problem is:
- in earlier versions of TB, the default implementation of a custom flot contained
```js
self.getSettingsSchema = function() {
    return TbFlot.settingsSchema('graph');
}
```
(and similar for getDataKeySettingsSchema).
- TbFlot.settingsSchema() was removed somewhere around 3.4.0. In the sources, there is a function `flotDefaultSettings()` in `ui-ngx/src/app/modules/home/components/widget/lib/settings/chart/flot-widget-settings.component.ts` but I can't find a way to call it from a component.
- If your self.getSettingsSchema returns {} or false or null, default settings will be used... probably (see `ui-ngx/src/app/shared/models/widget.models.ts`, class `WidgetSettingsComponent`, function `set settings`). However, the Advanced tab in widget settings does not show.
- To make the tab appear, the settingsSchema has to satisfy `displayAdvanced()` in `ui-ngx/src/app/modules/home/components/widget/widget-config.component.ts`, i.e. has to contain a key named "schema".

I can't find what or where should be put into getSettingsSchema().schema, or why it's not there. Minification of the sources does not help... I'm getting funny errors 30 levels deep in functions ...Gt(Ht(za(ws(Wo(pu(render(createSchema(u(G(... if I try to just add `schema: true`.

Does anyone have any idea how to proceed?
